### PR TITLE
Reland #18804

### PR DESCRIPTION
**Pull request by makslevental**
Reland https://github.com/iree-org/iree/pull/18804 including https://github.com/iree-org/iree/compare/main...ScottTodd:iree:ireegpu-api-fixes but also with (currently) a hack for exposing symbols in `compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp`.

TODO
- [x] decide/find the right way to expose symbols in `IREEGPUAttrs.cpp` on windows (alternatively move those C APIs)

EDIT: moved C wrappers to `compiler/src/iree/compiler/API/Internal/IREEGPUDialectCAPI.cpp`.

win CI job: https://github.com/iree-org/iree/actions/runs/11411731188

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/CMakeLists.txt [hunk 1]
```diff
@@ -40,6 +40,9 @@ iree_cc_library(
     "IREEGPUInterfaces.cpp"
     "IREEGPUOps.cpp"
     "TilingInterfaceImpl.cpp"
+  DEFINES
+    $<$<PLATFORM_ID:Windows>:MLIR_CAPI_ENABLE_WINDOWS_DLL_DECLSPEC>
+    $<$<PLATFORM_ID:Windows>:MLIR_CAPI_BUILDING_LIBRARY>
```
Comment (makslevental) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/CMakeLists.txt]: This works (at least locally - [gonna run win CI](https://github.com/iree-org/iree/actions/runs/11411230524) 🤞) but I doubt it's the "correct" thing to do.

File: compiler/src/iree/compiler/API/CMakeLists.txt [hunk 2]
```diff
@@ -82,6 +82,8 @@ set(_EXPORT_OBJECT_LIBS
   iree_compiler_API_Internal_IREEOptToolEntryPoint.objects
   iree_compiler_API_Internal_IREEReduceToolEntryPoint.objects
   iree_compiler_API_Internal_LLDToolEntryPoint.objects
+  # doesn't work because this target isn't known yet...
+  # iree_compiler_Codegen_Dialect_GPU_IR_IREEGPUDialect.objects
```
Comment (makslevental) [inline compiler/src/iree/compiler/API/CMakeLists.txt]: This doesn't work - maybe could work - but do we want to add `IREEGPUDialect` here?

**Review by ScottTodd — APPROVED** [review]
Cool! Looks about right to me, and builds successfully on my system.

**Comment by makslevental** [comment]
> Cool! Looks about right to me, and builds successfully on my system.

cool. gonna wait for win ci to finish/succeed just to be safe

**Review by kuhar — APPROVED** [review]

**Pull request merged**